packages/result/src/lib/Option.ts

```diff
@@ -1,50 +1,935 @@
+		return this.isSomeAnd((inner) => inner === value);
+	}
+
+	/**
+	 * Zips self with another `Option`.
+	 *
+	 * If self is `Some(s)` and `other` is `Some(o)`, this method returns `Some([s, o])`. Otherwise, `None` is returned.
+	 * @param other The option to zip self with.
+	 *
+	 * @example
+	 * ```typescript
+	 * const x = some(1);
+	 * const y = some('hi');
+	 * const z = none;
+	 *
+	 * assert.equal(x.zip(y), some([1, 'hi']));
+	 * assert.equal(x.zip(z), none);
+	 * ```
+	 *
+	 * @see {@link https://doc.rust-lang.org/std/option/enum.Option.html#method.zip}
+	 */
+	public zip<OtherValue, OtherExists extends boolean>(
+		other: Option<OtherValue, OtherExists>
+	): Option<[T, OtherValue], If<Exists, OtherExists, false>> {
+		// @ts-expect-error Complex types
+		return this.isSome() && other.isSome() ? some([this[ValueProperty], other[ValueProperty]] as [T, OtherValue]) : none;
+	}
+
+	/**
+	 * Zips self and another `Option` with function `f`.
+	 *
+	 * If self is `Some(s)` and other is `Some(o)`, this method returns `Some(f(s, o))`. Otherwise, `None` is returned.
+	 * @param other The option to zip self with.
+	 * @param f The function that computes the returned value.
+	 *
+	 * @example
+	 * ```typescript
+	 * class Point {
+	 *   public readonly x: number;
+	 *   public readonly y: number;
+	 *
+	 *   public constructor(x: number, y: number) {
+	 *     this.x = x;
+	 *     this.y = y;
+	 *   }
+	 * }
+	 *
+	 * const x = some(17.5);
+	 * const y = some(42.7);
+	 *
+	 * assert.equal(x.zipWith(y, (s, o) => new Point(s, o)), some(new Point(17.5, 42.7)));
+	 * assert.equal(x.zipWith(none, (s, o) => new Point(s, o)), none);
+	 * ```
+	 *
+	 * @see {@link https://doc.rust-lang.org/std/option/enum.Option.html#method.zip_with}
+	 */
+	public zipWith<OtherValue, OtherExists extends boolean, ReturnValue>(
+		other: Option<OtherValue, OtherExists>,
+		f: (s: T, o: OtherValue) => ReturnValue
+	): Option<ReturnValue, If<Exists, OtherExists, false>> {
+		// @ts-expect-error Complex types
+		return this.isSome() && other.isSome() ? some(f(this[ValueProperty], other[ValueProperty])) : none;
+	}
+
+	/**
+	 * Unzips an option containing a tuple of two options.
+	 *
+	 * If self is `Some((a, b))` this method returns `[Some(a), Some(b)]`. Otherwise, `[None, None]` is returned.
+	 *
+	 * @example
+	 * ```typescript
+	 * const x: Option<[number, string]> = some([1, 'hi']);
+	 * assert.equal(x.unzip(), [some(1), some('hi')]);
+	 * ```
+	 * @example
+	 * ```typescript
+	 * const x: Option<[number, string]> = none;
+	 * assert.equal(x.unzip(), [none, none]);
+	 * ```
+	 *
+	 * @see {@link https://doc.rust-lang.org/std/option/enum.Option.html#method.unzip}
+	 */
+	public unzip<Value0, Value1, Exists extends boolean>(
+		this: Option<readonly [Value0, Value1], Exists>
+	): [Option<Value0, Exists>, Option<Value1, Exists>] {
+		// @ts-expect-error Complex types
+		return this.match({
+			some: ([value0, value1]) => [some(value0), some(value1)],
```

Comment:
you don't check anywhere that `this` is a tuple/zipped up option

Reply:
Nor did the original code:
https://github.com/sapphiredev/utilities/blob/809c41cef869bb9095cd630a416981be6870e83a/packages/result/src/lib/Option/Some.ts#L142-L145

It's "checked" with the `this` guard, which hides this method altogether if you try to use it on an incompatible `Option`.